javasource/databaseconnector/impl/JdbcConnector.java

```diff
@@ -28,87 +30,145 @@
-				else
-					obj.setValue(context, name, convertedValue);
-			};
-
-			columns.forEach(setMemberValue);
-			logNode.trace("Instantiated object: " + obj);
-			return obj;
-		};
-
-		return executeQuery(jdbcUrl, userName, password, metaObject, sql).map(toMendixObject);
-	}
-
-	private Function<Object, Object> toSuitableValue(final PrimitiveType type) {
-		return v -> type == PrimitiveType.Binary ? new ByteArrayInputStream((byte[]) v) : v;
-	}
-
-	private Stream<Map<String, Optional<Object>>> executeQuery(final String jdbcUrl, final String userName,
-			final String password, final IMetaObject metaObject, final String sql) throws SQLException {
-		logNode.trace(String.format("executeQuery: %s, %s, %s", jdbcUrl, userName, sql));
-
-		try (Connection connection = connectionManager.getConnection(jdbcUrl, userName, password);
-				PreparedStatement preparedStatement = connection.prepareStatement(sql);
-				ResultSet resultSet = preparedStatement.executeQuery()) {
-			ResultSetReader resultSetReader = new ResultSetReader(resultSet, metaObject);
-
-			return resultSetReader.readAll().stream();
-		}
-	}
-
-	public long executeStatement(final String jdbcUrl, final String userName, final String password, final String sql)
-			throws SQLException {
-		logNode.trace(String.format("executeStatement: %s, %s, %s", jdbcUrl, userName, sql));
-
-		try (Connection connection = connectionManager.getConnection(jdbcUrl, userName, password);
-				PreparedStatement preparedStatement = connection.prepareStatement(sql)) {
-			return preparedStatement.executeUpdate();
-		}
-	}
+    private final ILogNode logNode;
+    private final ObjectInstantiator objectInstantiator;
+    private final ConnectionManager connectionManager;
+
+    public JdbcConnector(final ILogNode logNode, final ObjectInstantiator objectInstantiator,
+                         final ConnectionManager connectionManager) {
+        this.logNode = logNode;
+        this.objectInstantiator = objectInstantiator;
+        this.connectionManager = connectionManager;
+    }
+
+    public JdbcConnector(final ILogNode logNode) {
+        this(logNode, new ObjectInstantiatorImpl(), ConnectionManagerSingleton.getInstance());
+    }
+
+    public Stream<IMendixObject> executeQuery(final String jdbcUrl, final String userName, final String password,
+                                              final IMetaObject metaObject, final String sql, final IContext context) throws SQLException {
+        return executeQuery(jdbcUrl, userName, password, metaObject, sql, new ArrayList<>()).map(CreateMendixObjectConverter(context, metaObject));
+    }
+
+    public Stream<IMendixObject> executeQuery(final String jdbcUrl, final String userName, final String password,
+                                              final IMetaObject metaObject, final IStringTemplate sql, final IContext context) throws SQLException {
+
+        String queryTemplate = sql.replacePlaceholders((placeholderString, index) -> "?");
```

Comment:
Hm, what if we have a query like `Select Name from Users where FirstName like '%{1}%' or LastName like '%{1}%'`? Two placeholders referring to a single parameter.

Reply:
That's a good point. AFAIK named parameters are not supported in JDBC so we can't reuse parameters. This query will cause a runtime error. I'll try to see if there are any workarounds.

Reply:
Did a little investigations. Plain jdbc connector does not support named parameters. We can either use spring-jdbc instead of the plain jdbc connector or make the parameters list valid by duplicating the values in correct order. I'll give the latter a try.

Reply:
Done.